File: nanobot/agent/psb_tags.py
"""PSB 回复标签解析与剥离。"""

from __future__ import annotations

import re
from typing import Any

_PSB_TAG_RE = re.compile(
    r'<psb:(timeline|expression|face|fade)\b([^>]*?)/?>',
    re.IGNORECASE,
)
_ATTR_RE = re.compile(r"""(\w+)\s*=\s*(?:"([^"]*)"|'([^']*)')""")


def _parse_attrs(attr_text: str) -> dict[str, str]:
    attrs: dict[str, str] = {}
    for match in _ATTR_RE.finditer(attr_text):
        key = match.group(1)
        value = match.group(2) if match.group(2) is not None else match.group(3)
        attrs[key] = value or ""
    return attrs


def parse_psb_tag_actions(text: str) -> list[dict[str, Any]]:
    """从文本中解析 PSB 标签为运行时动作列表。"""
    actions: list[dict[str, Any]] = []
    for match in _PSB_TAG_RE.finditer(text or ""):
        tag_type = match.group(1).lower()
        attrs = _parse_attrs(match.group(2))
        if tag_type in {"timeline", "expression"}:
            name = attrs.get("name") or attrs.get("label") or ""
            if name:
                actions.append({"type": tag_type, "payload": {"name": name}})
        elif tag_type in {"face", "fade"}:
            var_name = attrs.get("var") or attrs.get("name") or ""
            raw_value = attrs.get("value")
            if var_name and raw_value is not None and raw_value != "":
                actions.append(
                    {
                        "type": tag_type,
                        "payload": {"var": var_name, "value": raw_value},
                    }
                )
    return actions


def strip_psb_tags(text: str) -> str:
    """去掉 PSB 特殊标签，用于展示或 TTS。"""
    cleaned = _PSB_TAG_RE.sub("", text or "")
    return re.sub(r"[ \t]{2,}", " ", cleaned).strip()

File: electron/src/psb/psb-tags.ts
/** PSB 回复标签解析（与 nanobot/web/psb/psb-tags.js 逻辑一致）。 */

export type PsbTagAction = {
  type: string;
  payload: Record<string, unknown>;
};

const PSB_TAG_RE = /<psb:(timeline|expression|face|fade)\b([^>]*?)\/?>/gi;
const ATTR_RE = /(\w+)\s*=\s*(?:"([^"]*)"|'([^']*)')/g;

function parseAttrs(attrText: string): Record<string, string> {
  const attrs: Record<string, string> = {};
  for (const match of attrText.matchAll(ATTR_RE)) {
    const key = match[1];
    const value = match[2] ?? match[3] ?? "";
    attrs[key] = value;
  }
  return attrs;
}

function tagToAction(type: string, attrs: Record<string, string>): PsbTagAction | null {
  const normalized = type.toLowerCase();
  if (normalized === "timeline" || normalized === "expression") {
    const name = attrs.name || attrs.label || "";
    if (!name) return null;
    return { type: normalized, payload: { name } };
  }
  if (normalized === "face" || normalized === "fade") {
    const varName = attrs.var || attrs.name || "";
    const rawValue = attrs.value;
    if (!varName || rawValue === undefined || rawValue === "") return null;
    return { type: normalized, payload: { var: varName, value: rawValue } };
  }
  return null;
}

/** 从 assistant 文本解析 PSB 标签动作（按出现顺序）。 */
export function parsePsbTags(text: string): PsbTagAction[] {
  const actions: PsbTagAction[] = [];
  const source = String(text || "");
  for (const match of source.matchAll(PSB_TAG_RE)) {
    const action = tagToAction(match[1], parseAttrs(match[2]));
    if (action) actions.push(action);
  }
  return actions;
}

/** 去掉 PSB 标签，用于聊天展示。 */
export function stripPsbTags(text: string): string {
  return String(text || "")
    .replace(PSB_TAG_RE, "")
    .replace(/[ \t]{2,}/g, " ")
    .trim();
}

/** 按配置决定是否隐藏 PSB 标签。 */
export function formatAssistantContentForDisplay(
  content: string,
  showResponseTags: boolean | undefined,
): string {
  if (showResponseTags) return content;
  return stripPsbTags(content);
}
